libgalois/src/PropertyGraph.cpp

```diff
@@ -530,3 +530,222 @@ katana::SortNodesByDegree(katana::PropertyGraph* pg) {
+  }
+
+  uint64_t num_nodes_symmetric = topology.num_nodes();
+  uint64_t num_edges_symmetric =
+      (*out_indices_symmetric)[num_nodes_symmetric - 1];
+
+  // Temp LargeArray for computing new destination positions
+  out_indices_symmetric_temp[0] = 0;
+  katana::do_all(
+      katana::iterate(uint64_t{1}, topology.num_nodes()),
+      [&](uint64_t n) {
+        out_indices_symmetric_temp[n] = (*out_indices_symmetric)[n - 1];
+      },
+      katana::no_stats());
+
+  out_dests_symmetric->allocateInterleaved(num_edges_symmetric);
+
+  // Update graph topology with the original edges + reverse edges
+  katana::do_all(
+      katana::iterate(uint64_t{0}, topology.num_nodes()),
+      [&](uint64_t src) {
+        auto edges = topology.edges(src);
+        // e = start index into edge array for a particular node
+        auto e = edges.begin();
+
+        // get all outgoing edges (excluding self edges) of a particular
+        // node in the and add reverse edges.
+        while (e < edges.end()) {
+          // destination node
+          auto dest = topology.edge_dest(*e);
+
+          // Add original edge
+          auto e_new_src =
+              __sync_fetch_and_add(&(out_indices_symmetric_temp[src]), 1);
+          (*out_dests_symmetric)[e_new_src] = dest;
+
+          // Do not add reverse edge for self-loops
+          if (dest != src) {
+            // Add reverse edge
+            auto e_new_dst =
+                __sync_fetch_and_add(&(out_indices_symmetric_temp[dest]), 1);
+            (*out_dests_symmetric)[e_new_dst] = src;
+            // TODO(gill) copy edge data to "new" array
+          }
+          e++;
+        }
+      },
+      katana::no_stats());
+
+  auto numeric_array_out_indices =
+      std::make_shared<arrow::NumericArray<arrow::UInt64Type>>(
+          static_cast<int64_t>(num_nodes_symmetric),
+          arrow::MutableBuffer::Wrap(
+              out_indices_symmetric->data(), num_nodes_symmetric));
+
+  auto numeric_array_out_dests =
+      std::make_shared<arrow::NumericArray<arrow::UInt32Type>>(
+          static_cast<int64_t>(num_edges_symmetric),
+          arrow::MutableBuffer::Wrap(
+              out_dests_symmetric->data(), num_edges_symmetric));
+
+  if (auto r = pfg_symmetric->SetTopology(katana::GraphTopology{
+          .out_indices = std::move(numeric_array_out_indices),
+          .out_dests = std::move(numeric_array_out_dests),
+      });
+      !r) {
+    return r.error();
+  }
+
+  return std::unique_ptr<PropertyGraph>(std::move(pfg_symmetric));
+}
+
+katana::Result<std::unique_ptr<katana::PropertyGraph>>
+katana::CreateTransposeGraph(katana::PropertyGraph* pg) {
+  const GraphTopology& topology = pg->topology();
+  auto pfg_tranpose = std::make_unique<katana::PropertyGraph>();
+  if (topology.num_nodes() == 0) {
+    return std::unique_ptr<PropertyGraph>(std::move(pfg_tranpose));
+  }
+
+  // New transpose graph topology
+  katana::LargeArray<uint64_t>* out_indices_transpose =
```

Comment:
ditto.

I feel like there is some possibility for reuse among `CreateTransposeGraph` and `CreateSymmetricGraph`. This might become more pressing when supporting properties as well.